pkg/cri/resource-manager/policy/builtin/dynamic-pools/dyp.go

```diff
@@ -0,0 +1,938 @@
+	cIDs := []string{}
+	for _, ctrIDs := range dp.PodIDs {
+		cIDs = append(cIDs, ctrIDs...)
+	}
+	return cIDs
+}
+
+// ContainerCount returns the number of containers in a dynamicPool.
+func (dp DynamicPool) ContainerCount() int {
+	count := 0
+	for _, ctrIDs := range dp.PodIDs {
+		count += len(ctrIDs)
+	}
+	return count
+}
+
+// AvailMilliCpus returns the number of CPUs in a dynamicPool.
+func (dp DynamicPool) AvailMilliCpus() int {
+	return dp.Cpus.Size() * 1000
+}
+
+// updateRealCpuUsed returns cpu utilization of a dynamicPool.
+func (p *dynamicPools) updateRealCpuUsed(dp *DynamicPool) (int, error) {
+	if dp.Cpus.Size() == 0 {
+		log.Info("dynamic pool %s cpuset is 0", dp.Def.Name)
+		return 0, nil
+	}
+	cpuInfo, _ := getCpuUtilization(time.Duration(time.Second))
+	cpus := dp.Cpus.ToSlice()
+	var sum float64
+	for i := 0; i < len(cpus); i++ {
+		sum += cpuInfo[cpus[i]]
+	}
+	res := int(sum / float64(100))
+
+	log.Info("dynamic pool %s cpuset: %s,  cpu utilization: %d", dp.Def.Name, dp.Cpus, res)
+	return res, nil
+}
+
+// calculateAllPoolWeights returns weights of all dynamicPools and the sum of weights.
+// Use dynamicPool's cpu utilization as its weight.
+func (p *dynamicPools) calculateAllPoolWeights() (map[*DynamicPool]int, int, error) {
+	weight := make(map[*DynamicPool]int)
+	sumWeight := 0
+	for _, dp := range p.dynamicPools {
+		if dp.Def.Name == "reserved" {
+			continue
+		}
+		realCpuUsed, err := p.updateRealCpuUsed(dp)
+		if err != nil {
+			return weight, sumWeight, dynamicPoolsError("The actual cpu usage of the dynamic pool %s cannot be obtained: %w",
+				dp.PrettyName(), err)
+		}
+		weight[dp] = realCpuUsed
+		sumWeight += weight[dp]
+		log.Info("dyanmic pool: %s, realCpuUsed: %d, weight: %d", dp, realCpuUsed, weight[dp])
+	}
+	log.Info("sum weight: %d", sumWeight)
+	return weight, sumWeight, nil
+}
+
+// calculateAllPoolRequests returns the sum of the requests of containers in each dynamicPool and remaining free cpu.
+// remainFree = allowed cpu - reserved cpu - sum(requests of containers in each dynamicPool)
+func (p *dynamicPools) calculateAllPoolRequests() (map[*DynamicPool]int, int) {
+	requestCpu := make(map[*DynamicPool]int)
+	remainFree := p.allowed.Difference(p.reserved).Size()
+	for _, dp := range p.dynamicPools {
+		if dp.Def.Name == "reserved" {
+			continue
+		}
+		requestCpu[dp] = (p.requestedMinMilliCpus(dp) + 999) / 1000
+		remainFree -= requestCpu[dp]
+		log.Info("dyanmic pool %s request cpu %d", dp, requestCpu[dp])
+	}
+	log.Info("sum remain free cpu %d", remainFree)
+	return requestCpu, remainFree
+}
+
+func (p *dynamicPools) containerPinPool(dp *DynamicPool) {
+	dp.Mems = p.closestMems(dp.Cpus)
+	for _, cID := range dp.ContainerIDs() {
+		if c, ok := p.cch.LookupContainer(cID); ok {
+			p.pinCpuMem(c, dp.Cpus, dp.Mems)
+		}
+	}
+}
+
+// updatePoolCpuset updates the cpuset of the dynamicPools.
+func (p *dynamicPools) updatePoolCpuset() error {
+	requestCpu, remainFree := p.calculateAllPoolRequests()
```

Comment:
A situations to consider and include in e2e tests: none of the containers in a dynamic pool request any CPUs. (The containers may be BestEffort or they may have only resource limit.) It looks to me that in that case the container might get an empty cpuset, which will mean "can run on any CPU".